<!-- Simulation of Husky UGV with GMapping and Frontier Exploration -->

<launch>
	<arg name="world"
		default="$(find husky_gazebo)/launch/playpen.launch" />
	<arg name="frontier_exploration" default="false" />
	<arg name="explore_lite" default="false" />
	<arg name="autonomy_cmd_vel_topic" default="/cmd_vel" />
	<arg name="goal_plugin" default="statemachine::CalculateGoalState" />
	<arg name="navigation_plugin"
		default="statemachine::RonaNavigationState" />
	<arg name="mapping_plugin" default="statemachine::MappingState" />
	<arg name="waypoint_routines" default="['Reversing']" />
	<arg name="rviz" default="false" />

	<include
		file="$(find statemachine_additions)/launch/gazebo.launch">
		<arg name="world_file" value="$(arg world)" />
	</include>

	<include file="$(find statemachine)/launch/statemachine.launch">
		<arg name="update_frequency" value="20" />
		<arg name="teleop_topic" value="/joy/cmd_vel" />
		<arg name="robot_frame" value="base_footprint" />
		<arg name="autonomy_cmd_vel_topic"
			value="$(arg autonomy_cmd_vel_topic)" />
		<arg name="calculate_goal_plugin"
			value="statemachine::CalculateGoalState" />
		<arg name="navigation_plugin"
			value="statemachine::NavigationState" />
		<arg name="mapping_plugin" value="$(arg mapping_plugin)" />
		<arg name="waypoint_routines" value="$(arg waypoint_routines)" />
	</include>

	<node pkg="statemachine_additions" type="bootUpNode"
		name="bootUpNode" output="screen" />

	<node pkg="gmapping" type="slam_gmapping" name="slam_gmapping"
		output="screen">
		<param name="base_frame" value="base_footprint" />
	</node>

	<node pkg="move_base" type="move_base" respawn="false"
		name="move_base" output="screen">
		<rosparam
			file="$(find statemachine_additions)/config/gazebo_sim/costmap_common.yaml"
			command="load" ns="global_costmap" />
		<rosparam
			file="$(find statemachine_additions)/config/gazebo_sim/costmap_common.yaml"
			command="load" ns="local_costmap" />
		<rosparam
			file="$(find statemachine_additions)/config/gazebo_sim/costmap_local.yaml"
			command="load" />
		<rosparam
			file="$(find statemachine_additions)/config/gazebo_sim/costmap_global.yaml"
			command="load" />
		<rosparam
			file="$(find statemachine_additions)/config/gazebo_sim/planner.yaml"
			command="load" />
		<remap from="/odom" to="/odometry/filtered" />
		<remap from="/cmd_vel" to="$(arg autonomy_cmd_vel_topic)" />
	</node>

	<!-- Explore Lite -->
	<group if="$(arg explore_lite)">
		<node pkg="explore_lite" type="explore" respawn="false"
			name="explore">
			<param name="robot_base_frame" value="base_footprint" />
			<param name="costmap_topic" value="map" />
			<param name="costmap_updates_topic" value="map_updates" />
			<param name="visualize" value="true" />
			<param name="planner_frequency" value="0.33" />
			<param name="progress_timeout" value="30.0" />
			<param name="potential_scale" value="3.0" />
			<param name="orientation_scale" value="0.0" />
			<param name="gain_scale" value="1.0" />
			<param name="transform_tolerance" value="0.3" />
			<param name="min_frontier_size" value="0.75" />
			<remap from="/move_base" to="/frontier_move_base" />
		</node>

		<node
			if="$(eval goal_plugin == 'statemachine::CalculateGoalState')"
			pkg="statemachine_additions" type="frontiersMarkerToPointcloudNode"
			name="frontiersMarkerToPointcloudNode" />

		<node
			if="$(eval goal_plugin == 'statemachine::CalculateGoalState')"
			pkg="statemachine_additions" type="goalInterceptNode"
			name="goalInterceptNode" output="screen" />
	</group>

	<!-- Frontier Exploration -->
	<group if="$(arg frontier_exploration)">

		<node if="$(arg frontier_exploration)" pkg="frontier_exploration"
			type="explore_client" name="explore_client" output="screen" />

		<node if="$(arg frontier_exploration)" pkg="frontier_exploration"
			type="explore_server" name="explore_server" output="screen">
			<param name="frequency" value="1.0" />
			<!-- Should be less than sensor range -->
			<param name="goal_aliasing" value="2.0" />
			<remap from="/move_base" to="/frontier_move_base" />
			<rosparam
				file="$(find statemachine_additions)/config/gazebo_sim/costmap_common.yaml"
				command="load" ns="explore_costmap" />
			<rosparam
				file="$(find statemachine_additions)/config/gazebo_sim/costmap_exploration.yaml"
				command="load" ns="explore_costmap" />
		</node>

		<node pkg="statemachine_additions" type="unboundExplorationNode"
			name="unboundExplorationNode" output="screen">
			<param name="unbound" value="true" />
			<param name="delay" value="10" />
			<param name="xoffset" value="7.5" />
			<param name="yoffset" value="7.5" />
			<param name="xdim" value="15.0" />
			<param name="ydim" value="15.0" />
			<param name="goalx" value="5.0" />
			<param name="goaly" value="5.0" />
		</node>

		<node
			if="$(eval goal_plugin == 'statemachine::CalculateGoalState')"
			pkg="statemachine_additions" type="goalInterceptNode"
			name="goalInterceptNode" output="screen" />
	</group>


	<!-- RViz fÃ¼r Exploration -->
	<node if="$(arg rviz)" pkg="rviz" type="rviz" name="rviz"
		args="-d $(find statemachine_additions)/config/gazebo_sim/gazebo_simulation.rviz" />
	<include if="$(arg rviz)"
		file="$(find statemachine_rviz_plugins)/launch/statemachine_rviz_plugins.launch">
		<arg name="update_frequency" value="20" />
	</include>
</launch>
